### ISSUE TITLE
multi target not supported issue in ClassNLLCriterion.lua

### ISSUE THREAD

**Issue by jbansal** [torch/cutorch#227]
Hi All,
Lately I have update my CUDA to cuda 7.5 and CUNN v3.0.

When I try to use ClassNLLCriterion, I am getting following error.

/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: multi-target not supported at /tmp/luarocks_cunn-scm-1-8806/cunn/ClassNLLCriterion.cu:118
stack traceback:
	[C]: in function 'ClassNLLCriterion_updateOutput'
	/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: in function 'forward'
	Main.lua:163: in function 'opfunc'
	/usr/local/share/lua/5.1/optim/sgd.lua:43: in function 'sgd'
	Main.lua:186: in main chunk
	[C]: in function 'dofile'
	[string "_RESULT={dofile "Main.lua" }"]:1: in main chunk
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/trepl/init.lua:630: in function 'repl'
	/usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00406260	

Please help me what could be the reason for this.
Regards,
Jitendra Bansal

**Comment by soumith** [comment]
https://github.com/torch/cunn/blob/master/ClassNLLCriterion.cu#L117-L119 you are passing >1D tensor

**Comment by jbansal** [comment]
Hi soumith,
Thanks for reply.
Yes i am passing tensor in 2dim array ex. Target[10][1].

It is working on cuda 7.0 and below version. I got this issue when i updated to 7.5

**Comment by fmassa** [comment]
@jbansal I don't think it's because of Cuda 7.5. It's simply because you updated your `cunn`, and it's behaviour changed. See https://github.com/torch/cunn/issues/127

**Comment by jbansal** [comment]
Hi fmssa,
I have seen the example mentioned in torch/cunn#127
When I run first print, I got following error

 print(m:forward(a,t2)) 
/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: multi-target not supported at /tmp/luarocks_nn-scm-1-2705/nn/generic/ClassNLLCriterion.c:22
stack traceback:
	[C]: in function 'ClassNLLCriterion_updateOutput'
	/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: in function 'forward'
	[string "print(m:forward(a,t2)) "]:1: in main chunk
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/trepl/init.lua:648: in function 'repl'
	/usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00406260	

After moving all into CUDA and then do print, I got following error

print(m:forward(a,t2)) 
/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: multi-target not supported at /tmp/luarocks_cunn-scm-1-9234/cunn/ClassNLLCriterion.cu:118
stack traceback:
	[C]: in function 'ClassNLLCriterion_updateOutput'
	/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: in function 'forward'
	[string "print(m:forward(a,t2)) "]:1: in main chunk
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/trepl/init.lua:648: in function 'repl'
	/usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00406260	

Please help me to identify the issue.

Regards,
Jitendra Bansal

**Comment by soumith** [comment]
Pass in 1D tensor, your problem will be fixed.

**Comment by jbansal** [comment]
Hi Soumith,
Thanks for this suggestion. I will try this.

**Comment by mattdns100689** [comment]
Hi there. I am had the same problem just now and passed a 1D tensor which worked! Is it not possible to push a batch > 1 in size through? Or am I missing something obvious here.

Many thanks,
Matt

**Comment by fmassa** [comment]
@mattdns100689 you can just provide one target per batch element, multi-target is not supported anymore.
So if your batch size is 128 and you have 10 classes, your target should be a `Tensor` of size `128`, with values ranging from 1 to 10.

**Comment by mattdns100689** [comment]
@fmassa thanks for clearing that up!

**Comment by fguney** [comment]
I have a question: how does this affect CrossEntropyCriterion?
In the documentation:

> The input given through a forward() is expected to contain scores for each class: input has to be a 1D Tensor of size n. This criterion expect a class index (1 to the number of class) as target when calling forward(input, target) and backward(input, target).

First, if it's a score per class, I expect to be able to input something like batch size x number of classes.
Second, if the input is scores, target should also be scores instead of class index, right?

**Comment by fmassa** [comment]
@fguney `CrossEntropyCriterion` behaves similarly to `ClassNLLCriterion` in this aspect.
So yes, the input should be `batch_size x number_of_classes`, but the target will be only `batch_size`, with elements in the range `1, 2, ..., number_of_classes`.

**Comment by sunshineatnoon** [comment]
What if my output and target are both n*n images tensor ？In this case，can I use ClassNLLCriterion to measure the difference between output and target？

**Comment by soumith** [comment]
@sunshineatnoon no, in that case, you can use cudnn.SpatialClassNLLCriterion

**Comment by fpthink** [comment]
http://pytorch.org/docs/nn.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss
Shape:
Input: (N,C)(N,C) where C = number of classes
Target: (N)(N) where each value is 0 <= targets[i] <= C-1

**Comment by brando90** [comment]
why is it so picky weather its Nx1 or N,?

**Comment by mehdirostami** [comment]
I had K-class classification problem. This fixed my problem:

I first defined:

target = Variable(torch.from_numpy(t_train[:batch_size]).long(), requires_grad = False)

loss = torch.nn.CrossEntropyLoss()
Then, in the training step, I defined:
...
y_pred = mlp_instance(x_data)
l = loss(y_pred, target)
...
where y_pred is a  batch_size by K tensor but target is a tensor (simply a vector) of size batch_size; note that target isn't one-hot tensor with shape: batch_size by K.

**Comment by dmorris0** [comment]
Mehdirostami,

Can you explain how your code converts one-hot to index values -- assuming it does this?

Thanks

> 
> 
> I had K-class classification problem. This fixed my problem:
> 
> I first defined:
> 
> target = Variable(torch.from_numpy(t_train[:batch_size]).long(), requires_grad = False)
> 
> loss = torch.nn.CrossEntropyLoss()
> Then, in the training step, I defined:
> ...
> y_pred = mlp_instance(x_data)
> l = loss(y_pred, target)
> ...
> where y_pred is a batch_size by K tensor but target is a tensor (simply a vector) of size batch_size; note that target isn't one-hot tensor with shape: batch_size by K.

**Comment by mehdirostami** [comment]
One way is this:

http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html

You can also write your own code:. For example:
Let t be target with K categories: 0, 1, ..., K-1

list_t = [(t == k).reshape(-1, 1) for k in range(K)]
one_hot_t = np.concatenate(list_t, axis=1)+0)

**Comment by fmassa** [comment]
this is btw in the wrong repo (this is Lua-Torch), but an even simpler way of converting one-hot to index is as follows:
```python
one_hot.argmax(dim=1)
```
